library/pkcs7.c

```diff
@@ -169,6 +169,91 @@ static int pkcs7_get_digest_algorithm_set(unsigned char **p,
     return 0;
 }
 
+/**
+ * SignedAttributes ::= SET SIZE (1..MAX) OF Attribute
+ *
+ * Attribute ::= SEQUENCE {
+ *      attrType OBJECT IDENTIFIER,
+ *      attrValues SET OF AttributeValue }
+ **/
+static int pkcs7_get_signed_attrs(unsigned char **p,
+                                  unsigned char *end,
+                                  mbedtls_x509_buf *attrs_raw,
+                                  mbedtls_asn1_named_data *attrs)
+{
+    mbedtls_asn1_named_data *cur = attrs;
+    unsigned char *seq_end;
+    size_t len;
+    int ret = 0;
+
+    attrs_raw->p = *p;
+    attrs_raw->len = 0;
+
+    ret = mbedtls_asn1_get_tag(p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+                               | MBEDTLS_ASN1_CONTEXT_SPECIFIC);
+    if (ret != 0) {
+        *p = attrs_raw->p;
+        attrs_raw->p = NULL;
+        return 0;
+    }
```

Comment:
Should this return 0 if `ret != MBEDTLS_ERR_ASN1_UNEXPECTED_TAG`?   The value of `*p` is undefined in this case, but I see that it is explicitly restored from `attrs_raw->p`, and the error should be detected by `pkcs7_get_digest_algorithm()`.  I think this code is correct in the context of its only caller, but it is confusing and better as follows:

```suggestion
    if (ret != 0) {
        *p = attrs_raw->p;
        attrs_raw->p = NULL;
        return ret == MBEDTLS_ERR_ASN1_UNEXPECTED_TAG ? 0 : ret;
    }
```